AsyncDisplayKit/Private/ASDisplayNode+UIViewBridge.mm

```diff
@@ -235,36 +234,63 @@ - (CGRect)frame
+      if (nodeLoaded && !pendingState.hasChanges) {
+        [[ASPendingStateController sharedInstance] registerNode:self];
+      }
+      pendingState.frame = rect;
+    }
+  } else {
+    if (nodeLoaded && isMainThread) {
+      // We do have to set frame directly, and we're on main thread with a loaded node.
+      // Just set the frame on the view.
+      // NOTE: Frame is only defined when transform is identity because we explicitly diverge from CALayer behavior and define frame without transform.
 #if DEBUG
-    // Checking if the transform is identity is expensive, so disable when unnecessary. We have assertions on in Release, so DEBUG is the only way I know of.
-    ASDisplayNodeAssert(CATransform3DIsIdentity(self.transform), @"-[ASDisplayNode setFrame:] - self.transform must be identity in order to set the frame property.  (From Apple's UIView documentation: If the transform property is not the identity transform, the value of this property is undefined and therefore should be ignored.)");
+      // Checking if the transform is identity is expensive, so disable when unnecessary. We have assertions on in Release, so DEBUG is the only way I know of.
+      ASDisplayNodeAssert(CATransform3DIsIdentity(self.transform), @"-[ASDisplayNode setFrame:] - self.transform must be identity in order to set the frame property.  (From Apple's UIView documentation: If the transform property is not the identity transform, the value of this property is undefined and therefore should be ignored.)");
 #endif
-    _view.frame = rect;
-  } else if (!nodeLoaded || (isMainThread && !setFrameDirectly)) {
-    /**
-     * Sets a new frame to this node by changing its bounds and position. This method can be safely called even if
-     * the transform is a non-identity transform, because bounds and position can be set instead of frame.
-     * This is NOT called for synchronous nodes (wrapping regular views), which may rely on a [UIView setFrame:] call.
-     * A notable example of the latter is UITableView, which won't resize its internal container if only layer bounds are set.
-     */
-    CGRect bounds;
-    CGPoint position;
-    ASBoundsAndPositionForFrame(rect, self.bounds.origin, self.anchorPoint, &bounds, &position);
-
-    self.bounds = bounds;
-    self.position = position;
-  } else if (nodeLoaded && !isMainThread) {
-    if (!_pendingViewState.hasChanges) {
-      [ASPendingStateController.sharedInstance registerNode:self];
+      _view.frame = rect;
+    } else {
+      // We do have to set frame directly, but either the node isn't loaded or we're on a non-main thread.
+      // Set the frame on the pending state, and it'll call setFrame: when applied.
+      _ASPendingState *pendingState = ASDisplayNodeGetPendingState(self);
+      if (nodeLoaded && !pendingState.hasChanges) {
+        [[ASPendingStateController sharedInstance] registerNode:self];
+      }
+      pendingState.frame = rect;
```

Comment:
Here I went with 4 total cases when I might have been able to get away with 3 (combining cases 1 and 3) but I found it damaged readability because it was unclear what possibility space that combined case covered. This way the comments very clearly explain each condition and why we're responding the way we do.